### DISCUSSION TITLE
Cannot kill the server if I attempt to run it along with another asyncio process.

### DISCUSSION THREAD

**Discussion by ashleney** [fastapi/fastapi#6716] [Questions]
When I want to run a fastapi server with uvicorn along with my discord bot I can do it simply by letting the server serve as a task, but this makes it impossible to kill both the server and the bot. I feel the problem here is that the server is waiting for the loop to close but the bot is preventing it or perhaps the other way around. All in all the process simply cannot be killed with `ctrl+c`.  

I've made this simple example with specifically a discord bot. (The discord bot runs what could be simplified to `loop.run_until_complete(self.start(...))`) Once the loop is started I let the server serve and everything works just fine except for of course the killing.

```py
import asyncio
from discord import Client
from fastapi import FastAPI
import uvicorn

app = FastAPI()

class MyClient(Client):
    async def start(self, *args, **kwargs) -> None:
        config = uvicorn.Config(app)
        server = uvicorn.Server(config)
        asyncio.create_task(server.serve())
        
        await super().start(*args, **kwargs)

client = MyClient()
client.run(TOKEN)
```
The script simply hangs at this after pressing `ctrl+c`
```
INFO:     Started server process [10312]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete. 
INFO:     Finished server process [10312]
```

* OS: Windows 10 
* FastAPI Version: 0.67.0
* Uvicorn Version:  0.13.4
* Python version: 3.9.0

I know that many people who open issues regarding discord bots are the same people who just picked up python yesterday but I ask you to please not think too harshly of me for not coming up with a better example.

**Comment by Kludex** [comment]
Hey @thesadru ! 

The issue here is that `discord` adds a signal handler for SIGINT (`CTRL` + `C`) and `uvicorn` does the same just after (so the `discord` handler is replaced by the `uvicorn` one). Uvicorn finishes successfully, but the event loop doesn't stop (as `discord` wants). 

You can run `uvicorn` in another process to solve the issue. 

This issue would be more appropriate either on `uvicorn` or `discord`, jfyk.

EDIT: You can try to readd the discord signal after the `create_task`. I'm away from my computer now.

**Comment by abhishek0220** [comment]
Hey @thesadru 
You can modify your `MyClient` to something like this
```python
class MyClient(Client):
    async def start(self, *args, **kwargs) -> None:  
        asyncio.create_task(self.start_uvicorn())      
        await super().start(*args, **kwargs)

    async def start_uvicorn(self):
        config = uvicorn.Config(app)
        server = uvicorn.Server(config)
        await server.serve()
        await self.close()
```

**Comment by ashleney** [comment]
Thanks, that solved the problem.
